bika/lims/catalog/indexers/analysisrequest.py

```diff
@@ -28,3 +28,41 @@ def assigned_state(instance):
             return 'unassigned'
 
     return 'assigned'
+
+
+@indexer(IAnalysisRequest)
+def listing_searchable_text(instance):
+    """ Indexes values of desired fields for searches in listing view. All the
+    field names added to 'plain_text_fields' will be available to search by
+    wildcards.
+    Please choose the searchable fields carefully and add only fields that
+    can be useful to search by. For example, there is no need to add 'SampleId'
+    since 'getId' of AR already contains that value. Nor 'ClientTitle' because
+    AR's are/can be filtered by client in Clients' 'AR Listing View'
+    :return: values of the fields defined as a string
+    """
+    entries = []
+    plain_text_fields = ('getId', 'getContactFullName', 'getSampleTypeTitle',
+                         'getSamplePointTitle', 'getCreatorFullName',
+                         'getProfilesTitle', 'getStorageLocationTitle',
+                         'getClientOrderNumber', 'getClientReference',
+                         'getClientSampleID', 'getTemplateTitle', )
+
+    # Concatenate plain text fields as they are
+    for field_name in plain_text_fields:
+        try:
+            value = api.safe_getattr(instance, field_name)
+        except:
+            logger.error("{} has no attribute called '{}' ".format(
+                            repr(instance), field_name))
+            continue
+
+        if not value:
+            continue
+        if isinstance(value, list):
+            value = " ".join(value)
+
+        entries.append(value)
+
+    # Concatenate all strings to one text blob
+    return " ".join(entries)
```

Comment:
Hi @nihadness, is it possible to provide a pluggable index function which works the same like https://github.com/senaite/senaite.core/blob/master/bika/lims/browser/bika_listing.py#L1209 ?

The explicit approach here lacks custom fields and therefore won't find any remarks, dates, review_title etc.

Does this search works as well case insensitive?

Reply:
@ramonski , it could be possible to add all the fields but it is not a good approach to save all values twice. I think beside multiplying memory of AR Listing Catalog by two, it would also slow down the queries. Default Fields such as remarks, dates, review_title and etc. can be added to the list as well, if it makes sense from functional point of view.
Yes you are right that when new field is added and is searchable, then it must be added to the list as well. But this how also Plone's 'SearchableTextFields' work. When a new field is added, it must be defined in Searchable Text Field from ZMI or from the code. So maybe we can sacrifice at this point.
Yes, everything saved in the field is lower case and search terms are always lowered as well.

Reply:
I think the memory wouldn't fall into weight compared to the explicit approach here.

Using the fields from the metadata would make this whole machinery quite flexible and usable for all catalogs. New searchable values could be added w/o code change by simply adding a metadata column to the catalog.

Otherwise the behavior of the search in AR listing behaves differently to all the other listings which use the metadata_search method.

In fact I would need to remove the listing index which comes in from the upgrade step for my customers after this release, because they are searching for columns not listed here...

Reply:
Sometimes, optimizing for performance comes with a cost. With @nihadness ' approach, the cost is that indeed, you'll need to specifically define for which fields you want a `portal_type` object to be searchable in lists. The amount and complexity of data one would expect from an Analysis Requests list and from other lists like Methods, Analysis Services, etc. is incomparable. 
`metadata_search` will work just fine for most of the lists and you will probably not notice any difference with `ng3_search`in terms of performance if the number of objects is <10k. But there are a these few lists (ARs, Worksheets and Samples) that will grow indefinitely and will easily reach >10k. At this point, I truly believe that adding the `NG3Index` is inalienable, no matter if the listing behaves differently to all other listings. 

On the other hand, enabling all lists to use `NG3Index` approach would be an error imo. Mostly because `metadata_search` will just work fine in most cases and adding a `NG3Index` to all them will come with non-justifiable costs: more code and more data in database, but without any benefit.

Disabling `NG3Index` for a given customer is not harder than adding a client-specific metadata for a given catalog. Just remove the index from the catalog and `metadata_search` will substitute `ng3_search` without the need of further actions.

If we agree that some lists (only 3 I can think of) might behave differently, then there is no reason to add all metadata in `NG3Index`, better to just add those that are meaningful, but allow others to be added in the indexer if necessary. In case we want those indexer to behave different in lab-specific add-ons, or don't like those strings hard-coded there, then we can always override that indexer or make use of adapters to get the fields transparently.

Reply:
Is see the need for the index @xispa, no question. I only don't understand why we not simply put the output of [`metadata_to_searchable_text `](https://github.com/senaite/senaite.core/blob/master/bika/lims/browser/bika_listing.py#L1209) into this indexer, so I can use it as it is.

Reply:
Yes, I totally agree with you on that it will require code addition when adding a new index and also it is good to search for string everywhere from functional point of view, because it is obviously desired to get all the objects containing search term in any field. It would work like `ctrl+f`.

But by adding all metadata values to that index, we lose the point of 'Metadata & Index' logic. The whole point of Indexes in any DB is to save them ordered by each Index. Basically, every time you add a new Index, it costs more than adding just a metadata or a column (in the terms of memory and speed while querying). That's why we should always carefully avoid which metadata columns we do not want to search by, and not add them as Indexes to our DB or Catalog. It is also the reason why Searchable Text Indexes from Plone lets you add values from Indexes to that SearchableTextIndex, not metadata values.

I mean I don't like the idea of using metadata values in queries to DB. I think it would be more understandable, to get all the Indexes and add all index values to that new field, instead of adding all metadata values...

Reply:
Didn't see the last comments, so my previous comment is kind of outdated now.
@ramonski , I didn't push the final changes yet, I will definitely use `metadata_to_searchable_text` method inside the `listing_searchable_text` function to save values. I am just waiting for the final decision from this discussion to start working on _Requested Review Changes_.

Reply:
Ok, maybe I'm missing here something...

What is the difference between

- Concatenating the values of the explicit getters into a single string and put that into the index
- Getting all metadata, concatenate the strings together and put it then to the index?

Ok, maybe we have some bytes more in the string, but does it really count? We could maybe skip the dates, so it would be even more similar in my opinion.

Reply:
well, I think just having more bytes in the strings is acceptable at some point . But I also think-considering values are saved in the list inside `NG3Index`, adding more elements (strings) to that list, will slow down the queries.
If you think it wouldn't make a big difference, then we can simply add all the metadata columns there and see.

Reply:
Searching and finding contents in the sysstem is crucial in my opinion for user acceptance.

Having a quick search, which does not find the expected results, will lower this acceptance and I's rather accept more memory/time consumption to make provide a good search facility than to lose user acceptance.

Also users expect to wait after they entered a search value. We should rather improve performance on pure display sites and listings.

Anyhow, please consider that in your coding @nihadness and thanks for your replies so far.

Reply:
of course, I always try to consider things we talk about in PR's, very helpful for self-improvement :)
So, I will just modify the method and add all the metadata values to that index.
Welcome and thank you and Jordi too!

Reply:
Just thinking (for another PR) that maybe could be nice to create a function like `serch_by_term(catalog, portal_type, search_term, base_query=None)` in `senaite.core.api` and move all related logic (`metadata_search` and `ngx3_search`) there?
btw, these discussions are one of the reasons I love you all! ;)